React/Modules/RCTUIManager.m

```diff
@@ -723,6 +727,20 @@ - (void)_amendPendingUIBlocksWithStylePropagationUpdateForShadowView:(RCTShadowV
   }
 }
 
+- (void)addUIManagerObserver:(id<RCTUIManagerObserver>)observer
+{
+  [_uiManagerObserversLock lock];
```

Comment:
The order of operations that currently touch the lock (addUIManagerObserver, removeUIManagerObserver, the code within flushUIBlocks) should still happen in the same order with dispatch_async, which keeps the order of blocks (it uses FIFO). I believe that "otherwise the observer could be added after the UI operations are completed" is about as equally likely with the locks as with dispatch_async.